File: src/com/ptby/dynamicreturntypeplugin/scripting/ScriptReplacementExecutor.java
// require(url:'https://scripting.dev.java.net', jar:'groovy-engine.jar')
package com.ptby.dynamicreturntypeplugin.scripting;

import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.ptby.dynamicreturntypeplugin.scripting.api.ExecutingScriptApi;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

public class ScriptReplacementExecutor {

    public static final String SCRIPT_LANGUAGE_JAVASCRIPT = "JavaScript";
    public static final String SCRIPT_LANGUAGE_GROOVY = "groovy";

    private final PhpCallReferenceInfo phpCallReferenceInfo;
    private final CallableScriptConfiguration callableScriptConfiguration;
    private final Invocable invocable;
    private ScriptSignatureParser scriptSignatureParser;


    public ScriptReplacementExecutor( String scriptLanguage,
                                      PhpCallReferenceInfo phpCallReferenceInfo,
                                      CallableScriptConfiguration callableScriptConfiguration ) throws ScriptException {
        this.phpCallReferenceInfo = phpCallReferenceInfo;
        this.callableScriptConfiguration = callableScriptConfiguration;

        ExecutingScriptApi executingScriptApi = new ExecutingScriptApi( callableScriptConfiguration.getScriptFileLocation() );

        ScriptEngineManager manager = new ScriptEngineManager();
        ScriptEngine engine = manager.getEngineByName( scriptLanguage );
        if ( engine == null ) {
            throw new ScriptException(
                    "Script engine '" + scriptLanguage + "' was not created. Relevant jar may not be in classpath."
            );
        }

        engine.put( "api", executingScriptApi );

        engine.eval( callableScriptConfiguration.getScriptCode() );
        invocable = ( Invocable ) engine;
        scriptSignatureParser = new ScriptSignatureParser();
    }


    public String executeAndReplace( String currentValue ) {
        ParsedSignature parsedSignature = scriptSignatureParser.parseSignature( currentValue );
        if ( parsedSignature == null ) {
            return "";
        }

        try {
            Object result = invocable.invokeFunction(
                    callableScriptConfiguration.getScriptCall(),
                    parsedSignature.getNamespace(),
                    parsedSignature.getReturnClassName()
            );

            return parsedSignature.getPrefix() + String.valueOf( result );
        } catch ( ScriptException e ) {
            String message = "Error executing " + callableScriptConfiguration.getScriptCall() + " in " + callableScriptConfiguration
                    .getScriptFileLocation() + "\n" + e.getMessage();
            Notifications.Bus.notify( createWarningNotification( message ) );
        } catch ( NoSuchMethodException e ) {
            String message = "No such method " + callableScriptConfiguration.getScriptCall() + " in " + callableScriptConfiguration
                    .getScriptFileLocation() + "\n" + e.getMessage();

            Notifications.Bus.notify( createWarningNotification( message ) );
        }

        return "";
    }


    private Notification createWarningNotification( String message ) {
        return new Notification(
                "DynamicReturnTypePlugin",
                "Script file error",
                message,
                NotificationType.WARNING
        );
    }
}

File: src/com/ptby/dynamicreturntypeplugin/scripting/ScriptReplacementExecutor.kt
// require(url:'https://scripting.dev.java.net', jar:'groovy-engine.jar')
package com.ptby.dynamicreturntypeplugin.scripting

import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.project.Project
import com.jetbrains.php.PhpIndex
import com.ptby.dynamicreturntypeplugin.index.ClassConstantWalker
import com.ptby.dynamicreturntypeplugin.scripting.api.ExecutingScriptApi
import com.ptby.dynamicreturntypeplugin.signature_extension.startsWithClassConstantPrefix
import com.ptby.dynamicreturntypeplugin.signature_extension.startsWithMethodCallPrefix
import javax.script.Invocable
import javax.script.ScriptException

class ScriptReplacementExecutor @Throws(ScriptException::class) constructor(customScriptEngineFactory: CustomScriptEngineFactory,
                                                                            val callableScriptConfiguration: CallableScriptConfiguration) {
    private val invocable: Invocable
    private val scriptSignatureParser: ScriptSignatureParser
    private val classConstantWalker = ClassConstantWalker()

    init {
        val executingScriptApi = ExecutingScriptApi(this)
        val engine = customScriptEngineFactory.getEngine()
        engine.eval(callableScriptConfiguration.code)
        engine.put("api", executingScriptApi)

        invocable = engine as Invocable
        scriptSignatureParser = ScriptSignatureParser()
    }


    fun executeAndReplace(project: Project, currentValue: String): String {
        val parsedSignature = parseSignature(project, currentValue)
                ?: return ""

        try {
            val result = invocable.invokeFunction(
                    callableScriptConfiguration.call,
                    parsedSignature.namespace,
                    parsedSignature.returnClassName
            )
            return parsedSignature.prefix + result
        } catch (e: ScriptException) {
            val message = "Error executing " + callableScriptConfiguration.call + " in " +
                    callableScriptConfiguration.fileLocation + "\n" + e.message

            Notifications.Bus.notify(createWarningNotification(message))
        } catch (e: NoSuchMethodException) {
            val message = "No such method " + callableScriptConfiguration.call + " in " +
                    callableScriptConfiguration.fileLocation + "\n" + e.message

            Notifications.Bus.notify(createWarningNotification(message))
        }

        return ""
    }

    private fun parseSignature(project: Project, currentValue: String): ParsedSignature? {

        val signatureToParse =
                if ( currentValue.startsWithMethodCallPrefix()) {
                    val index = PhpIndex.getInstance(project)
                    val bySignature = index.getBySignature(currentValue)
                    if ( bySignature.size > 0){
                        bySignature.first().type.toString()
                    }else {
                        ""
                    }
                } else if ( !currentValue.startsWithClassConstantPrefix() ) {
                    currentValue
                } else {
                    classConstantWalker.walkThroughConstants(project, currentValue) ?:
                            currentValue
                }

        if( signatureToParse == "" ){
            return null
        }

        return scriptSignatureParser.parseSignature(signatureToParse)
    }


    private fun createWarningNotification(message: String): Notification {
        return Notification("DynamicReturnTypePlugin", "Script file error", message, NotificationType.WARNING)
    }

    companion object {
        const public val SCRIPT_LANGUAGE_JAVASCRIPT: String = "JavaScript"
        const public val SCRIPT_LANGUAGE_GROOVY: String = "groovy"
    }
}
